In [93]:
# Vector3 Class
from dataclasses import dataclass
import math

@dataclass(frozen=True)
class Vector3:
    x: float
    y: float
    z: float

    # Vector addition
    def __add__(self, other: "Vector3") -> "Vector3":
        return Vector3(
            self.x + other.x,
            self.y + other.y,
            self.z + other.z
        )

    # Vector subtraction
    def __sub__(self, other: "Vector3") -> "Vector3":
        return Vector3(
            self.x - other.x,
            self.y - other.y,
            self.z - other.z
        )

    # Scalar multiplication
    def __mul__(self, scalar: float) -> "Vector3":
        return Vector3(
            self.x * scalar,
            self.y * scalar,
            self.z * scalar
        )
    
    # Scalar division
    def __truediv__(self, scalar: float) -> "Vector3":
        return Vector3(
            self.x / scalar,
            self.y / scalar,
            self.z / scalar
        )

    __rmul__ = __mul__
    __rtruediv__ = __truediv__

    # Dot product
    def dot(self, other: "Vector3") -> float:
        return (
            self.x * other.x +
            self.y * other.y +
            self.z * other.z
        )

    # Cross product
    def cross(self, other: "Vector3") -> "Vector3":
        return Vector3(
            self.y * other.z - self.z * other.y,
            self.z * other.x - self.x * other.z,
            self.x * other.y - self.y * other.x
        )

    # Magnitude (length)
    def magnitude(self) -> float:
        return math.sqrt(self.dot(self))

    # Unit vector
    def unit_vector(self) -> "Vector3":
        mag = self.magnitude()
        if mag == 0:
            raise ValueError("Cannot normalize a zero vector")
        return self * (1 / mag)

In [94]:
# Semi Major Axis (a)
def compute_sma(mu:float, energy:float):
    return -mu/(2*energy)


In [95]:
# Eccentricity (e)
import math

def compute_ecc(energy: float, mu: float, h:float):
    return math.sqrt(1+(2*math.pow(h, 2)*energy)/(math.pow(mu, 2)))

In [96]:
# Inclination (i)
import math

def compute_inc(Z_: Vector3, h_: Vector3):
    Z_ = Z_.unit_vector()
    h_ = h_.unit_vector()
    return math.acos(Z_.dot(h_))


In [97]:
# RAAN (capital omega) Ω
import math

def compute_raan(Nx: float, Ny: float):
    return math.atan2(Ny, Nx)

In [98]:
# Argument of Perigeee (omega sub p) ωp
import math

def compute_argp(h_: Vector3, N_: Vector3, B_: Vector3):
    h_
    N_
    B_
    return math.atan2(h_.dot(N_.cross(B_)), N_.dot(B_))



In [99]:
# True Anomaly (nu) ν
import math

def compute_ta(v_: Vector3, r_: Vector3, B_: Vector3):
    cosTa = r_.dot(B_)/(r_.magnitude()*B_.magnitude())
    ta = math.acos(cosTa)
    if r_.dot(v_) < 0:
        ta = 2*math.pi - ta
    return ta

In [100]:
# Period (TP)
import math

def compute_period(a: float, mu: float):
    return 2*math.pi*math.sqrt(math.pow(a, 3)/mu)

In [101]:
# Apogee (r sub a) ra
def compute_apogee(a: float, e: float):
    return a*(1+e)

In [102]:
# Perigee (r sub p) rp
def compute_perigee(a: float, e: float):
    return a*(1-e)

In [103]:
# Energy (xi) ξ
import math

def compute_energy(v: float, r: float, mu: float):
    return math.pow(v, 2)/2 - mu/r

In [104]:
# Angular Momentum Vector (h_)
def compute_angular_momentum_vector(r: Vector3, v: Vector3):
    return r.cross(v)

In [105]:
# RAAN Vector (N_) Unitized
def compute_raan_vector(Z_: Vector3, h_: Vector3):
    Z_ = Z_.unit_vector()
    h_ = h_.unit_vector()
    return (Z_.cross(h_)/(Z_.cross(h_).magnitude())).unit_vector()

In [106]:
# Perigee Vector (B_) Unitized
def compute_perigee_vector(mu: float, r_: Vector3, v_: Vector3, h_: Vector3):
    return v_.cross(h_)-mu*(r_/r_.magnitude()).unit_vector()

In [107]:
# Keplerian Elements Class
import math

class KeplerianElements:
    """
    Computes classical Keplerian orbital elements from
    position and velocity state vectors.
    """

    mu_earth = 3.986004418e14  # m^3 / s^2

    def __init__(self, position: Vector3, velocity: Vector3):
        """
        Parameters
        ----------
        position : Vector3
            Position vector (meters)
        velocity : Vector3
            Velocity vector (m/s)
        """

        mu = self.mu_earth
        k_hat = Vector3(0, 0, 1)

        # Specific orbital energy
        self.xi = compute_energy(velocity.magnitude(), position.magnitude(), mu)

        # Semi-major axis
        self.a = compute_sma(mu, self.xi)

        # Angular momentum vector
        self.h_ = compute_angular_momentum_vector(position, velocity)

        # Eccentricity
        self.ecc = compute_ecc(self.xi, mu, self.h_.magnitude())

        # Inclination
        self.inc = compute_inc(k_hat, self.h_)
        self.inc_deg = math.degrees(self.inc)

        # RAAN
        self.N_Unit = compute_raan_vector(k_hat, self.h_)
        self.raan = compute_raan(self.N_Unit.x, self.N_Unit.y)
        self.raan_deg = math.degrees(self.raan)

        # Argument of perigee
        self.B_Unit = compute_perigee_vector(
            mu,
            position,
            velocity,
            self.h_
        )
        self.argp = compute_argp(self.h_, self.N_Unit, self.B_Unit)
        self.argp_deg = math.degrees(self.argp)

        # True anomaly
        self.ta = compute_ta(velocity, position, self.B_Unit)
        self.ta_deg = math.degrees(self.ta)

        # Orbital period (elliptical only)
        self.period = compute_period(self.a, mu)

        # Apoapsis / periapsis
        self.apogee = compute_apogee(self.a, self.ecc)
        self.perigee = compute_perigee(self.a, self.ecc)


```^ Previous Lessons ^```

# Problem 1:

In [116]:
pos = Vector3(326151.080726, 6077471.251787, 2944583.918767)  # meters
vel = Vector3(-7455.178720, -482.482572, 1910.883434) # m/s

kepler_elements = KeplerianElements(pos, vel)
print("nu (true anomaly) is: " + str(kepler_elements.ta) + " rad")

nu (true anomaly) is: 0.533708002792791 rad


In [117]:
# Eccentric Anomaly (E)
import math

def compute_eccentric_anomaly(nu: float, e: float) -> float:
    return math.asin((math.sin(nu)*math.sqrt(1 - math.pow(e, 2))) / (1 + e*math.cos(nu)))


In [118]:
E_SV = compute_eccentric_anomaly(kepler_elements.ta, kepler_elements.ecc)
print("E0 (eccentric anomaly) is: " + str(E) + " rad")

E0 (eccentric anomaly) is: 0.5286424011417852 rad


In [119]:
# Mean Anomaly (M)
import math

def compute_mean_anomaly(E: float, e: float) -> float:
    return E - e*math.sin(E)

In [120]:
M_SV = compute_mean_anomaly(E0, kepler_elements.ecc)
print("M0 (mean anomaly) is: " + str(M0) + " rad")

M0 (mean anomaly) is: 0.5235987858960514 rad


In [121]:
# Mean Motion (n)
import math

def compute_mean_motion(mu: float, a: float):
    # note a is semi major axis in meters
    return math.sqrt(mu/math.pow(a,3))

##### Time of flight from perigee to nu (true anomaly)
Initially:  
nu (true anomaly) is: 0.533708002792791 rad  
E0 (eccentric anomaly) is: 0.5286424011417852 rad  
M0 (mean anomaly) is: 0.5235987858960514 rad  
  
At Perigee:  
tp = 0  
Ep = 0  

General Time of Flight Equation:  
`t-t0 = kTP + (1/n)(E - e * sin(E)) - (1/n) * (E0 - e * sin(E0))`  
- T = Time of last perigee passage  
- t0 = Time from perigee to E0  
- t = Time from perigee to E  
- k = perigee passages between E0 and E  
- k * TP = number of full orbits from epoch  

Therefore  
t0 is 0 and k is 0 and E0 is 0 so  
`t = (1/n)(E - e*sin(E))`

In [122]:
n = compute_mean_motion(kepler_elements.mu_earth, kepler_elements.a)
print("n (mean motion) is: " + str(n) + " rad/sec") 

n (mean motion) is: 0.0011209657051213528 rad/sec


In [ ]:
t_delta = (1/n)*(E_SV-kepler_elements.ecc * math.sin(E_SV))
print("time delta between perigee and nu: " + str(t_delta) + " sec")

time delta between perigee and nu: 467.0961685124594 sec


##### Time of flight from nu to nu = 65deg
Initially:  
nu (true anomaly) is: 0.533708002792791 rad  
E0 (eccentric anomaly) is: 0.5286424011417852 rad  
M0 (mean anomaly) is: 0.5235987858960514 rad  
  
At Perigee:  
tp = 0  
Ep = 0  

General Time of Flight Equation:  
`t-t0 = kTP + (1/n)(E - e * sin(E)) - (1/n) * (E0 - e * sin(E0))`  
- T = Time of last perigee passage  
- t0 = Time from perigee to E0  
- t = Time from perigee to E  
- k = perigee passages between E0 and E  
- k * TP = number of full orbits from epoch  

Therefore  
t0 is 0 and k is 0 and E0 is 0 so  
`t = (1/n)(E - e*sin(E))`